File: m02_ros2_communication/m02_ros2_with_py/m02_ros2_with_py/add_two_nums_srv.py
# !/usr/bin/env python3

# ---------------------------- ROS2 Dependencies ------------------------------
import rclpy
from rclpy.node import Node
# ---------------------------- ROS2 Msg/Srv -----------------------------------
from example_interfaces.srv import AddTwoInts

# ---------------------------- Server Class Node ------------------------------
class SimpleAdderSrv(Node):
    """
    Simple class related with a server for addition

    Attributes
    ---
    srv : Service Object
        Server that attends request made in /add_two_ints and implements the
        adder callback

    Methods
    ---
    adder_callback(request, response)
        Function called when a request is received, read it and make the sum
    """
    def __init__(self):
        """
        Constructor that initialize the node and instance a server
        """
        super().__init__('add_two_nums_srv')
        self.srv = self.create_service(AddTwoInts, 'add_two_ints', 
                                       self.adder_callback)

    def adder_callback(self, request, response):
        """
        Callback to attend request for sum and respond with the total.

        Params
        ---
        request : Request object
            Contains the info of the two numbers to sums
        response : Response object
            For updating the value of the sum
        """
        response.sum = request.a + request.b 
        self.get_logger().info("Request: (%d , %d)" % (request.a, request.b))

# ---------------------------- Implementation ---------------------------------

def main():
    # Initialize ROS Client Library for Python
    rclpy.init()

    # Instance server for addition
    my_adder = SimpleAdderSrv()

    # Generate loop to give time to accept incoming requests
    rclpy.spin(my_adder)

    # Destroy node and shut resources
    my_adder.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: m02_ros2_communication/m02_ros2_with_cpp/src/add_two_nums_srv.cpp
/**
 * File: add_two_nums_srv.cpp
 * 
 * Description: Simple service for adding two nums with the rclcpp
 * implementation for ROS2.
*/

// ---------------------------- CPP Standard Libraries -----------------------

#include <memory> // General utitilites to manage dynamic memory

// -------------------------------- rclcpp Headers ---------------------------

#include "rclcpp/rclcpp.hpp"  // Proper include for using rclcpp

#include "example_interfaces/srv/add_two_ints.hpp" // Include srv file 


// -------------------------------- Function Prototypes -----------------------

void add(const std::shared_ptr<example_interfaces::srv::AddTwoInts::Request> 
            request,
         std::shared_ptr<example_interfaces::srv::AddTwoInts::Response>
            response);

// ------------------------------- MAIN PROGRAM -------------------------------
int main(int argc, char **argv)
{
  // ROS CLient Library for C++ initialization
  rclcpp::init(argc, argv);

  // Initialize node called 'add_two_ints_server' by defining a shared pointer
  std::shared_ptr<rclcpp::Node> node = 
    rclcpp::Node::make_shared("add_two_ints_server");

  // Instance a service server object that hear to request of AddTwoInts by the 
  // service 'add_two_ints' and linking a callback for responses.
  rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr service =
    node->create_service<example_interfaces::srv::AddTwoInts>("add_two_ints", 
                                                              &add);
  // Log that service is ready to attend
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Ready to add two ints.");

  // Spin node to manage multiple requests
  rclcpp::spin(node);

  // Close node and shutdown library usage
  rclcpp::shutdown();
}

// -------------------------------- Function Definitions ----------------------

/**
 * Add callback that will manage incoming requests of two integers two give the
 * proper response and send it.
 * 
 * @param request Pointer to the request provided by client
 * @param response Pointer to response that ig going to be sent to the client.
*/
void add(const std::shared_ptr<example_interfaces::srv::AddTwoInts::Request> 
           request,
         std::shared_ptr<example_interfaces::srv::AddTwoInts::Response>
           response)
{
  // Update response by making the operation
  response->sum = request->a + request->b;

  // Log info of the process
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"),"Request received \na: %ld  b: %ld",
                request->a, request->b);
  RCLCPP_INFO(rclcpp::get_logger("rclcpp"),"Sendind response: [%ld]", 
              (long int)response->sum);
}